test/es12/optional-chaining.js

```diff
@@ -0,0 +1,223 @@
+            assert.throws(() => simpleObj.null(), TypeError);
+            assert.throws(() => simpleObj.undefined(), TypeError);
+
+            // With optional-chains
+            assert.isUndefined(simpleObj.nothing?.(), "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.null?.(), "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.undefined?.(), "OptChain should evaluated to 'undefined'");
+        }
+    },
+    {
+        name: "Simple properties before call",
+        body() {
+            // Verify normal behavior
+            assert.throws(() => simpleObj.nothing.something(), TypeError);
+            assert.throws(() => simpleObj.null.something(), TypeError);
+            assert.throws(() => simpleObj.undefined.something(), TypeError);
+
+            // With optional-chains
+            assert.isUndefined(simpleObj.nothing?.something(), "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.null?.something(), "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.undefined?.something(), "OptChain should evaluated to 'undefined'");
+        }
+    },
+    {
+        name: "Simple indexers before call",
+        body() {
+            // Verify normal behavior
+            assert.throws(() => simpleObj.nothing["something"], TypeError);
+            assert.throws(() => simpleObj.null["something"], TypeError);
+            assert.throws(() => simpleObj.undefined["something"], TypeError);
+
+            // With optional-chains
+            assert.isUndefined(simpleObj.nothing?.["something"](), "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.null?.["something"](), "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.undefined?.["something"](), "OptChain should evaluated to 'undefined'");
+        }
+    },
+    {
+        name: "Short-circuiting ignores indexer expression and method args",
+        body() {
+            let i = 0;
+
+            assert.isUndefined(simpleObj.nothing?.[i++]);
+            assert.isUndefined(simpleObj.null?.[i++]);
+            assert.isUndefined(simpleObj.undefined?.[i++]);
+
+            assert.isUndefined(simpleObj.nothing?.[i++]());
+            assert.isUndefined(simpleObj.null?.[i++]());
+            assert.isUndefined(simpleObj.undefined?.[i++]());
+
+            assert.isUndefined(simpleObj.nothing?.something(i++));
+            assert.isUndefined(simpleObj.null?.something(i++));
+            assert.isUndefined(simpleObj.undefined?.something(i++));
+
+            assert.strictEqual(0, i, "Indexer may never be evaluated");
+
+            // ToDo: How can I run async tests?
+            // simpleObj.nothing?.[await Promise.reject()];
+            // simpleObj.null?.[await Promise.reject()];
+            // simpleObj.undefined?.[await Promise.reject()];

```

Comment:
How can I run async tests?

Reply:
There are a few ways, have a look at test/es7/async-generator-functionality.js for one example.

Note, I'd recommend putting the async tests in a separate file - I think it's easier/better to manage them separately.